nova_vm/src/ecmascript/builtins/primitive_objects.rs

```diff
@@ -183,7 +190,212 @@ impl InternalSlots for PrimitiveObject {
+    ) -> JsResult<Value> {
+        if let Ok(string) = String::try_from(agent[self].data) {
+            if let Some(string_desc) = string.get_property_descriptor(agent, property_key) {
+                return Ok(string_desc.value.unwrap());
+            }
+        }
+
+        // 1. Return ? OrdinaryGet(O, P, Receiver).
+        match self.get_backing_object(agent) {
+            Some(backing_object) => {
+                ordinary_get(agent, backing_object.into_object(), property_key, receiver)
+            }
+            None => {
+                // a. Let parent be ? O.[[GetPrototypeOf]]().
+                let Some(parent) = self.internal_get_prototype_of(agent)? else {
+                    // b. If parent is null, return undefined.
+                    return Ok(Value::Undefined);
+                };
+
+                // c. Return ? parent.[[Get]](P, Receiver).
+                parent.internal_get(agent, property_key, receiver)
+            }
+        }
+    }
+
+    fn internal_set(
+        self,
+        agent: &mut Agent,
+        property_key: PropertyKey,
+        value: Value,
+        receiver: Value,
+    ) -> JsResult<bool> {
+        if let Ok(string) = String::try_from(agent[self].data) {
+            if string
+                .get_property_descriptor(agent, property_key)
+                .is_some()
+            {
+                return Ok(false);
+            }
+        }
+
+        // 1. Return ? OrdinarySet(O, P, V, Receiver).
+        let backing_object = self
+            .get_backing_object(agent)
+            .unwrap_or_else(|| self.create_backing_object(agent))
+            .into_object();
+        ordinary_set(agent, backing_object, property_key, value, receiver)
+    }
+
+    fn internal_delete(self, agent: &mut Agent, property_key: PropertyKey) -> JsResult<bool> {
+        if let Ok(string) = String::try_from(agent[self].data) {
+            if string
+                .get_property_descriptor(agent, property_key)
+                .is_some()
+            {
+                return Ok(false);
+            }
+        }
+
+        // 1. Return ? OrdinaryDelete(O, P).
+        match self.get_backing_object(agent) {
+            Some(backing_object) => {
+                ordinary_delete(agent, backing_object.into_object(), property_key)
+            }
+            None => Ok(true),
+        }
+    }
+
+    fn internal_own_property_keys(self, agent: &mut Agent) -> JsResult<Vec<PropertyKey>> {
+        if let Ok(string) = String::try_from(agent[self].data) {
+            let len = string.len(agent);
+            let mut keys = Vec::with_capacity(len + 1);
+
+            // Insert keys for every index into the string.
+            keys.extend(
+                (0..len)
+                    .map(|idx| PropertyKey::Integer(SmallInteger::try_from(idx as u64).unwrap())),
+            );
+
+            let backing_object_keys;
+            let (integer_keys, other_keys) = match self.get_backing_object(agent) {
```

Comment:
**praise:** Bravo, this is awesome! I would not have found such a neat solution to this, at ALL!